module/os/linux/spl/spl-kmem.c

```diff
@@ -245,7 +245,14 @@ spl_kmem_alloc_impl(size_t size, int flags, int node)
 				return (NULL);
 			}
 		} else {
-			if (flags & KM_VMEM) {
+			/*
+			 * We only use kmalloc when doing
+			 * kmem_alloc(KM_NOSLEEP). Otherwise we use kvmalloc,
+			 * which tries to get contiguous physical memory (fast,
+			 * like kmalloc) and falls back on using virtual memory
+			 * to stitch together pages (slow, like vmalloc).
+			 */
+			if ((flags & KM_VMEM) || !(flags & KM_NOSLEEP)) {
 				ptr = spl_kvmalloc(size, lflags);
```

Comment:
... but, if you could consider adding a comment that kvmalloc might implicitly sleep, it'd immediately help grokking the rest of the logic/comment.